app/slave/cluster_slave.py

```diff
@@ -74,24 +75,36 @@ def _start_heartbeat(self):
 
     def _run_heartbeat(self):
         try:
-            self._send_heartbeat_to_master()
-            self._heartbeat_failure_count = 0
+            response = self._send_heartbeat_to_master()
+
+            # A non HTTP 200 response indicates that the master is up, but does not recognize the slave.
+            # The slave tries to reconnect to master at this point.
+            # This is also counted as a heartbeat failure and the slave takes an action accordingly
+            if response.status_code != http.client.OK:
+                self._act_on_heartbeat_failure()
+                self._logger.info('Master responded with {}. Attempting to reconnect.'.format(response.status_code))
```

Comment:
This is on purpose. The slave may decide to die inside `_act_on_heartbeat_failure` because the heartbeat threshold is reached. Say, a case where master is returning 500s. We only want to reconnect only if that is not the case.